### PR TITLE
fix(react): signIn provider type

### LINKED ISSUE #2654: Next auth 4.0.0 sign in provider will only accept Email or Credentials
### Description 🐜

Currently in 4.0.0 sign in provider type is 
```
export type RedirectableProvider = "email" | "credentials"
```

which means any other provider will be seen as an error to typescript.  I believe changing it to a optional string will fix the issue. 

### Is this a bug in your own project?

Yes

### How to reproduce ☕️

Following line will throw type error when building as google is not part of the RedirectableProvider 
```
signin("google"); 
```

### Screenshots / Logs 📽

_No response_

### Environment 🖥

System:
    OS: Windows 10 10.0.19043
    CPU: (24) x64 AMD Ryzen 9 5900X 12-Core Processor
    Memory: 17.91 GB / 31.92 GB
  Binaries:
    Node: 14.17.2 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 7.21.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1023.0), Chromium (93.0.961.38)
    Internet Explorer: 11.0.19041.906
  npmPackages:
    @preact/compat:  0.0.4
    next: 11.1.0 => 11.1.0
    next-auth: 4.0.0-next.26 => 4.0.0-next.26

### Contributing 🙌🏽

Yes,  I am willing to help solve this bug in a PR

### PR DESCRIPTION
**Pull request by Patryks1**
## Reasoning 💡

Currently the signIn provider type will only accept email and credentials. This pr will accept any string instead. 

## Checklist 🧢

- [ ] Documentation
- [ ] Tests
- [x] Ready to be merged

## Affected issues 🎟
Fixes #2654

### CODE REVIEW

File: src/react/index.tsx [hunk 1]
```diff
@@ -172,7 +172,7 @@ export async function getProviders() {
 export async function signIn<
   P extends RedirectableProvider | undefined = undefined
 >(
-  provider?: RedirectableProvider,
+  provider?: string,
```
Comment (balazsorban44) [inline src/react/index.tsx:175]: This seems to destroy the purpose of RedirectableProvider.

What we would like here is to hint the user of any of the built-in provider types, but not restricting it to those.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you for jumping on this!

Some thoughts.

Ideally I would like that whenever someone types `signIn("` in their IDE, a list of suggestions come up to show all the providers we have built-in support for. This currently includes `credentials`, `email` and any of the OAuth providers' id.

Example:


Although we don't want to restrict the first param of `signIn` to these values in case someone creates a custom provider, with a unique id.

For this to work, I found the following workaround (taken from https://github.com/microsoft/TypeScript/issues/29729#issuecomment-832522611):

```ts
// https://github.com/microsoft/TypeScript/issues/29729#issuecomment-832522611
export type LiteralUnion<T extends U, U = string> =
  | T
  | (U & Record<never, never>)

export type RedirectableProviderType = "email"  |  "credentials"
export type BuiltInProviderType = LiteralUnion<RedirectableProviderType | OAuthProviderType>
```
OAuthProviderType should be a union of id strings from across our `src/providers/*` folder. We will have to generate this type, by iterating over that folder and use the file name, and put it in a file like `src/providers/oauth-types`.

Without `LiteralUnion`:


With `LiteralUnion` (while still getting the hints for the built-in providers.):

**Comment by Patryks1** [comment]
@balazsorban44 Any ideas on how you could possibly grab all the providers dynamically? I can only think of require.context/filesystem which might work in this scenario. however i dont think they will be done on compile which means they wont show up on intel sense. I could do it the dirty way and just add the providers by hand?

**Comment by balazsorban44** [comment]
I would use a node script to grab filenames.

For inspiration https://github.com/nextauthjs/docs/blob/main/scripts/generate-providers.js

**Comment by Patryks1** [comment]
> I would use a node script to grab filenames.
> 
> For inspiration https://github.com/nextauthjs/docs/blob/main/scripts/generate-providers.js
> 
> Then call this script when the build script is run

Nice one, Let me know if thats what you where looking for.

File: package.json [hunk 2]
```diff
@@ -145,7 +146,8 @@
       ".next",
       "dist",
       "/server",
-      "/react.js"
+      "/react.js",
+      "/scripts"
```
Comment (balazsorban44) [inline package.json:150]:
```suggestion
      "/react.js"
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
So I had some suggestions, and since I wanted to get this PR as quickly as possible, I thought I was committing to your branch, but instead, it ended up on a totally different branch...

Could you incorporate this commit with the suggested changes to your branch?: https://github.com/nextauthjs/next-auth/commit/2d8270c7668201da732db073a2a1817fe68d0355

It is mostly moving things around a bit, you have done nice work!

File: tsconfig.json [hunk 3]
```diff
@@ -25,6 +25,7 @@
     "**/tests",
     "**/__tests__",
     "config",
-    "www"
+    "www",
+    "./scripts/*.js"
```
[the remark this answers is not in the document]
Comment (Patryks1) [reply tsconfig.json]: done

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Patryks1) [reply package.json]: done

**Pull request merged**